gems/sorbet-runtime/test/types/virally_install_hooks.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+require_relative '../test_helper'
+
+class Opus::Types::Test::VirallyInstallHooksTest < Critic::Unit::UnitTest
+  it "does not give the hooks when including a non-viral module" do
+    m = Module.new do
+      extend T::Sig
+      sig {void}
+      def foo; end
+    end
+    assert_equal(true, T::Private::Methods.instance_variable_get(:@installed_hooks).include?(m))
+    c = Class.new do
+      include m
+    end
+    assert_equal(false, T::Private::Methods.instance_variable_get(:@installed_hooks).include?(c))
+  end
+
+  it "does not give the hooks when extending a non-viral module" do
+    m = Module.new do
+      extend T::Sig
+      sig {void}
+      def foo; end
+    end
+    assert_equal(true, T::Private::Methods.instance_variable_get(:@installed_hooks).include?(m))
+    c = Class.new do
+      extend m
+    end
+    assert_equal(false, T::Private::Methods.instance_variable_get(:@installed_hooks).include?(c))
+  end
+
+  it "gives the hooks when including a viral module" do
+    m = Module.new do
+      T::Private::Methods.instance_variable_get(:@virally_install_hooks) << self
```

Comment:
Is the point of this instance variable just to be able to write this test?

Reply:
Hm. The point of this instance variable, in my mind, was to make it so that a module which, for example, defines a final method, would get the new `included` and `extended` behavior.

However, one thing I failed to account for is that a module might not have any final methods on it at the time it has `install_hooks` called on it, but it could gain final methods later. I think that means instead of using this `@virally_install_hooks`, it might be better to do something like just call `mod.extend(VirallyInstallHooks)` whenever a final method is added to `mod`. (It appears to be safe to call `extend` multiple times.)

In a related vein, I'm wondering if it was a mistake to try to split this PR away from the final method checks. You left another comments about the tests I wrote, but I wonder if it would be better to test if things are working by seeing if the final checks are running when we want them to be. WDYT about closing this PR and folding it into #1181?

Reply:
I think it's fine to fold them into one if you want.